lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -235,4 +241,60 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
+# Collect text output into JSON. save_results() is called too early for this.
+sub finalize_results {
+    my ($self) = @_;
+
+    my $dir = $self->job->result_dir();
+    return 0 unless $dir;
+    my $file    = path($dir, "details-" . $self->name . ".json");
+    my $tmpfile = $file->sibling($file->basename . '.tmp');
+    return 0 unless -e $file;
+    my $errors = 0;
+    my $results;
+
+    eval {
+        $results = decode_json($file->slurp);
+        $results = {details => $results} if ref($results) eq 'ARRAY';
+
+        for my $step (@{$results->{details}}) {
+            next unless $step->{text};
+            my $txtfile = path($dir, $step->{text});
+            next unless -e $txtfile;
+            $step->{text_data} = decode('UTF-8', $txtfile->slurp);
+        }
+
+        $tmpfile->spurt(encode_json($results));
+
+        if (!rename $tmpfile->to_string, $file->to_string) {
+            die "finalize_results(" . $self->job->id . "): Cannot update details-" . $self->name . ".json. $!";
+        }
+    };
+
+    if ($@) {
+        log_error $@;
```

Comment:
It's bad practice to use `$@` more than once. There's a high risk that someone might add cleanup code before this line that accidentally changes the value.
```perl
if (my $err = $@) {
...
```